docs-2.0/1.introduction/3.nebula-graph-architecture/4.storage-service.md

```diff
@@ -1,11 +1,261 @@
+
+### Listener
+
+Listener is designed for **storage expansion**. It takes a long time for the newly added machines to synchronize data during expansion. If synchronizing as a follower, the availability of the entire cluster will decrease.
+
+Listener will be written into the WAL command. If the leader finds a command of `add learner` when writing the WAL, it will add the learner to its peers and mark it as a Listener. Listeners will not be included into the majorities, but logs will still be sent to them as usual. Listeners themselves will not initiate elections.
+
+Raft listener can write the data into Elasticsearch cluster after receiving them from Learner to implement full-text search. For more information, see [Deploy Raft Listener](../../4.deployment-and-installation/6.deploy-text-based-index/3.deploy-listener.md)。
+
+### Transfer Leadership
+
+Transfer leadership is extremely important for balance. When moving a partition from one machine to another, Nebula Graph first checks if the source is a leader. If so, it should be moved to another peer. After data migration is completed, it is important to [balance leader distribution](../../8.service-tuning/load-balance.md) again.
+
+When a transfer leadership command is committed, the leader will abandon its leadership and the followers will start a leader election.
+
+### Peer changes
+
+To avoid split-brain, when members in a Raft Group change, an intermediate state is required. In such a state, the majority of the old group and new group always have an overlap. Thus it prevents the old or new group from making decisions unilaterally. To make it even simpler, in his doctoral thesis Diego Ongaro suggests adding or removing a peer once to ensure the overlap between the majority of the new group and the old group . Nebula Graph‘s implementation also uses this approach, except that the way to add or remove a member is different. For details, please refer to addPeer/removePeer in the Raft Part class.
+
+## Differences with HDFS
+
+Storage Service is a Raft-based distributed architecture, which has certain differences with that of HDFS. For example:
+
+- Storage service ensures consistency through Raft. Usually, the number of its replicas are odd to elect a leader. However, DataNode used by HDFS ensures consistency through NameNode, which has no limit on the number of replicas.
+
+- In Storage service, only the replicas of the leader can read and write, while in HDFS all the replicas can do so.
+
+- In Storage service, the number of replicas needs to be determined when creating a space, since it cannot be changed afterwards. But in HDFS, the number of replicas can be changed freely.
+
+- Storage service can access to the file system directly, while the upper layer of HDFS (such as HBase) needs to access HDFS before the file system, using RPC more times.
```

Comment:
Storage services can access the file system directly. While the applications of HDFS (such as HBase) have to access HDFS before the file system, which requires more RPC times.